### ISSUE TITLE
IQueryable<T> filter by subquery gives "Item with Same Key has already been added”

### ISSUE THREAD

**Issue by Molinware** [nhibernate/nhibernate-core#2659]
Execute this query, gives the following exception:
```
statelessSession.Query<Sale>()
    .Select(x => new SaleDto
    {
        Id = x.Id,
        SaleDate = x.ListChildren.Max(y => y.Date)
    })
    .Where(x => x.SaleDate <= DateTime.Today && x.SaleDate >= DateTime.Today)
    .ToList();
```
Exception:
`System.ArgumentException: Item with Same Key has already been added.`

**Comment by hazzik** [comment]
Can you provide a stack trace? Just message isn't valuable

**Comment by Molinware** [comment]
[ERROR]
An item with the same key has already been added.

[StackTrace]
   em System.ThrowHelper.ThrowArgumentException(ExceptionResource resource)
   em System.Collections.Generic.Dictionary`2.Insert(TKey key, TValue value, Boolean add)
   em NHibernate.Linq.Visitors.HqlGeneratorExpressionVisitor.VisitBinaryExpression(BinaryExpression expression)
   em NHibernate.Linq.Visitors.HqlGeneratorExpressionVisitor.VisitExpression(Expression expression)
   em NHibernate.Linq.Visitors.HqlGeneratorExpressionVisitor.VisitBinaryExpression(BinaryExpression expression)
   em NHibernate.Linq.Visitors.HqlGeneratorExpressionVisitor.VisitExpression(Expression expression)
   em NHibernate.Linq.Visitors.QueryModelVisitor.VisitWhereClause(WhereClause whereClause, QueryModel queryModel, Int32 index)
   em Remotion.Linq.QueryModelVisitorBase.VisitBodyClauses(ObservableCollection`1 bodyClauses, QueryModel queryModel)
   em Remotion.Linq.QueryModelVisitorBase.VisitQueryModel(QueryModel queryModel)
   em NHibernate.Linq.Visitors.QueryModelVisitor.Visit()
   em NHibernate.Linq.Visitors.QueryModelVisitor.GenerateHqlQuery(QueryModel queryModel, VisitorParameters parameters, Boolean root, Nullable`1 rootReturnType)
   em NHibernate.Linq.NhLinqExpression.Translate(ISessionFactoryImplementor sessionFactory, Boolean filter)
   em NHibernate.Hql.Ast.ANTLR.ASTQueryTranslatorFactory.CreateQueryTranslators(IQueryExpression queryExpression, String collectionRole, Boolean shallow, IDictionary`2 filters, ISessionFactoryImplementor factory)
   em NHibernate.Engine.Query.QueryPlanCache.GetHQLQueryPlan(IQueryExpression queryExpression, Boolean shallow, IDictionary`2 enabledFilters)
   em NHibernate.Impl.AbstractSessionImpl.GetHQLQueryPlan(IQueryExpression queryExpression, Boolean shallow)
   em NHibernate.Impl.AbstractSessionImpl.CreateQuery(IQueryExpression queryExpression)
   em NHibernate.Linq.DefaultQueryProvider.PrepareQuery(Expression expression, IQuery& query)
   em NHibernate.Linq.DefaultQueryProvider.ExecuteList[TResult](Expression expression)
   em NHibernate.Linq.NhQueryable`1.System.Collections.Generic.IEnumerable<T>.GetEnumerator()
   em System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   em System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   em Genesis.BO.CondominioBo.Teste() na D:\Fontes\Genesis.NET\Genesis.BO\Database\CondominioBo.Partial.cs:linha 34
   em SyncInvokeTeste(Object , Object[] , Object[] )
   em System.ServiceModel.Dispatcher.SyncMethodInvoker.Invoke(Object instance, Object[] inputs, Object[]& outputs)
   em System.ServiceModel.Dispatcher.DispatchOperationRuntime.InvokeBegin(MessageRpc& rpc)
   em System.ServiceModel.Dispatcher.ImmutableDispatchRuntime.ProcessMessage5(MessageRpc& rpc)
   em System.ServiceModel.Dispatcher.MessageRpc.Process(Boolean isOperationContextSet)

**Comment by hazzik** [comment]
Thanks.

This is regression of #2491

**Comment by hazzik** [comment]
It throws an exception on `_notCastableExpressions.Add`

https://github.com/nhibernate/nhibernate-core/blob/2b1af8f990d38ca6c9878736a71df8e6250f555b/src/NHibernate/Linq/Visitors/HqlGeneratorExpressionVisitor.cs#L316-L322

**Comment by maca88** [comment]
I am unable to reproduce the issue on version `5.3.5` with the following test:
```
[Test]
public void CanCompareAggregateResult()
{
  var query = Sfi.OpenStatelessSession().Query<Customer>()
         .Select(o => new Test {Id = o.CustomerId, MaxDate = o.Orders.Max(l => l.OrderDate.Value)})
         .Where(o => o.MaxDate <= DateTime.Today && o.MaxDate >= DateTime.Today)
         .ToList();
}

private class Test
{
  public string Id { get; set; }

  public DateTime MaxDate { get; set; }
}
```

@Molinware can you please provide the definition of `SaleDto` and mapping for `Sale` class?

**Comment by Molinware** [comment]
I used a fake class to simplify the example. Even so I've now created the class **Sales** and same error ocurrs:

**Sales.hbm.xml**
```
<?xml version="1.0" encoding="utf-8" ?>
<hibernate-mapping xmlns="urn:nhibernate-mapping-2.2" assembly="Genesis.OR" namespace="Genesis.OR" default-lazy="false">
  <class name="Genesis.OR.Sales,Genesis.OR" table="SALES">
    <cache usage="read-write" />
    <id name="IdSales" column="ID_SALES" type="Int64" unsaved-value="0">
      <generator class="sequence" >
        <param name="sequence">GEN_SALES_ID</param>
      </generator>
    </id>
    <version name="Alteracao" unsaved-value="undefined" type="Timestamp"> <column name="ALTERACAO" index="SALES_IDX1" /> </version>
    <many-to-one name="IdEmpresa" column="ID_EMPRESA" class="Empresa" not-null="true" index="SALES_IDX2" />
    <property column="LANCAMENTO" type="DateTime" name="Lancamento" not-null="true" index="SALES_IDX3" />
    <bag name="ListaSalesChildren" inverse="true" cascade="all-delete-orphan" lazy="false" >
      <key column="ID_SALES" />
      <one-to-many class="Genesis.OR.SalesChildren,Genesis.OR" />
    </bag>
  </class>
</hibernate-mapping>
```

**SalesChildren.hbm.xml**
```
<?xml version="1.0" encoding="utf-8" ?>
<hibernate-mapping xmlns="urn:nhibernate-mapping-2.2" assembly="Genesis.OR" namespace="Genesis.OR" default-lazy="false">
  <class name="Genesis.OR.SalesChildren,Genesis.OR" table="SALES_CHILDREN">
    <cache usage="read-write" />
    <id name="IdSalesChildren" column="ID_SALES_CHILDREN" type="Int64" unsaved-value="0">
      <generator class="sequence" >
        <param name="sequence">GEN_SALES_CHILDREN_ID</param>
      </generator>
    </id>
    <version name="Alteracao" unsaved-value="undefined" type="Timestamp"> <column name="ALTERACAO" index="SALES_CHILDREN_IDX1" /> </version>
    <many-to-one name="IdSales" column="ID_SALES" class="Sales" not-null="true" index="SALES_CHILDREN_IDX2" />
    <property column="LANCAMENTO" type="Date" name="Lancamento" not-null="true" />
  </class>
</hibernate-mapping>
```

**SalesDto.cs**
```
public class SalesDto
{
    public long Id { get; set; }
    public DateTime? MaxDate { get; set; }
}
```

**Execute Query**
```
SessionManager.ObterNovaSessaoStateless()
    .Query<Sales>()
    .Select(
         x => new SalesDto()
         {
             Id = x.IdSales,
             MaxDate = x.ListaSalesChildren.Max(x => x.Lancamento)
          })
    .Where(x => x.MaxDate >= DateTime.Today && x.MaxDate <= DateTime.Today)
    .ToList();
```

**Comment by bahusoid** [comment]
You should also provide entity classes to simplify reproduction.

> Downgraded to version 5.3.2 is working

So please confirm that you downgraded from **5.3.5** to **5.3.2**. Also did you try versions in between 5.3.3 and 5.3.4 - same error?

**Comment by Molinware** [comment]
> You should also provide entity classes to simplify reproduction.

**Sales.cs**
```
namespace Genesis.OR
{
    using System;
    using System.Collections.Generic;
    using System.Runtime.Serialization;
    using Base;

    [DataContract(IsReference = true)]
    [Serializable]
    public partial class Sales
    {
        private IList<SalesChildren> listaSalesChildren;

        public Sales()
        {
            this.Alteracao = DateTime.Now;
            this.IdEmpresa = null;
            this.IdSales = 0;
            this.Lancamento = DateTime.Now;
        }

        [DataMember]
        public virtual DateTime Alteracao { get; set; }

        [DataMember]
        public virtual Empresa IdEmpresa { get; set; }

        [DataMember]
        public virtual long IdSales { get; set; }

        [DataMember]
        public virtual DateTime Lancamento { get; set; }

        [DataMember]
        public virtual IList<SalesChildren> ListaSalesChildren
        {
            get
            {
                if (this.listaSalesChildren == null)
                {
                    this.listaSalesChildren = new PersistentGenericBagLocal<SalesChildren>();
                }

                return this.listaSalesChildren;
            }

            set
            {
                if (value is Array)
                {
                    this.listaSalesChildren = new PersistentGenericBagLocal<SalesChildren>(value);
                    return;
                }

                this.listaSalesChildren = value;
            }
        }
    }
}
```
**SalesChildren.cs**
```
namespace Genesis.OR
{
    using System;
    using System.Runtime.Serialization;
    using Base;

    [DataContract(IsReference = true)]
    [Serializable]
    public partial class SalesChildren
    {
        public SalesChildren()
        {
            this.Alteracao = DateTime.Now;
            this.IdSales = null;
            this.IdSalesChildren = 0;
            this.Lancamento = DateTime.Now;
        }

        [DataMember]
        public virtual DateTime Alteracao { get; set; }

        [DataMember]
        public virtual Sales IdSales { get; set; }

        [DataMember]
        public virtual long IdSalesChildren { get; set; }

        [DataMember]
        public virtual DateTime Lancamento { get; set; }
    }
}
```

> So please confirm that you downgraded from **5.3.5** to **5.3.2**. Also did you try versions in between 5.3.3 and 5.3.4 - same error?

**5.3.5**: Error
**5.3.4**: Not Tested
**5.3.3**: Not Tested
**5.3.2**: Working

Tonight I can teste this the two versions in between.

**Comment by fredericDelaporte** [comment]
Fixed by #2664.
